x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -85,6 +82,31 @@ public IndicesPermission(Group... groups) {
         }
     }
 
+    private Automaton indexMatcherAutomaton(String... indices) {
+        final List<String> exactMatch = new ArrayList<>();
+        final List<String> patternMatch = new ArrayList<>();
+        for (String indexPattern : indices) {
+            if (isIndexPattern(indexPattern)) {
+                patternMatch.add(indexPattern);
+            } else {
+                exactMatch.add(indexPattern);
+            }
+        }
+        try {
+            final Automaton exactMatchAutomaton = Automatons.patterns(exactMatch);
```

Comment:
this drops the exact match optimization. Why do we do it this way?

Reply:
Because we need it for `testIndexMatch` that is needed for "has privilege". In "has privilege" the user can check an index pattern (not only a concrete index as in the authorization use case), and we test this with the automaton subset operation. However, the exact optimization only works for predicates. The code for this has been copied from `TransportHasPrivilegesAction` and I agree it amounts to a great deal of complexity in this class.

Yet, I think it's the right approach to have the optimization for the "authorize" use case, and a cache-able automaton for the "has privilege" use case (even though the optimization partly aims to bypass automaton generation in the first place). "has privilege" should be rare, and we cannot avoid automatons in this case.